src/bos_os_dir.ml

```diff
@@ -70,30 +70,80 @@ let fold_contents ?err ?dotfiles ?elements ?traverse f acc d =
 
 (* User and current working directory *)
 
+let user_from_passwd entry =
+  let home = entry.Unix.pw_dir in
+  match Fpath.of_string home with
+  | Ok p -> Ok p
+  | Error _ ->
+      Fmt.error_msg
+        "could not parse path (%a) from passwd entry"
+        String.dump home
+
 let user () =
   let debug err = Bos_log.debug (fun m -> m "OS.Dir.user: %s" err) in
-  let env_var_fallback () =
-    Result.bind (Bos_os_env.(parse "HOME" (some path) ~absent:None)) @@
+  let env_var var_name =
+    Result.bind (Bos_os_env.(parse var_name (some path) ~absent:None)) @@
     function
     | Some p -> Ok p
     | None -> Fmt.error_msg "cannot determine user home directory: \
-                            HOME environment variable is undefined"
+                            %s environment variable is undefined"
+                            var_name
   in
-  if Sys.os_type = "Win32" then env_var_fallback () else
-  try
-    let uid = Unix.getuid () in
-    let home = (Unix.getpwuid uid).Unix.pw_dir in
-    match Fpath.of_string home with
+  if Sys.os_type = "Win32" then
+    (* We should check USERPROFILE exclusively (without HOME) because otherwise it could break
+       Diskuv OCaml Windows users. See the discussion at https://github.com/dbuenzli/bos/pull/96.
```

Comment:
```suggestion
       native Windows users. See the discussion at https://github.com/dbuenzli/bos/pull/96.
```

Reply:
@jonahbeckford FYI: I changed it to "... break a Windows setup with MSVC++ or Mingw" for consistency.